New rule: Text nodes have minimal contrast (#833)
act-rules/act-rules.github.io: _rules/text-contrast-afw4f7.md

rule_type: atomic

description: |
  This rule checks that the highest possible contrast of every text character with its background meets the minimal contrast requirement

accessibility_requirements:
  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

acknowledgements:
  authors:
    - Brian Bors
    - Kasper Isager
    - Wilco Fiers
---

## Applicability

Any [visible](#visible) character in a [text node](https://dom.spec.whatwg.org/#text) that is a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled]()

> - Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled]()

Reviewer: The final condition is not clear. It looks like it is the semantic role that is disabled. I think it should be an "and"

- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) and is [disabled]()

**Note**: When the text color is the same as the background color, the [text node](https://dom.spec.whatwg.org/#text) is not [visible](#visible), and so it does not need to be tested for contrast.

## Expectation

> For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

Author: I don't think that works all that well. The way that would have to be worded is like this, which is rather repetitive:

> For each test target, either:
> - the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1, or
> - the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 3.0:1 for [larger scale text](#large-scale-text), or
> - the test target is part of a [text node](https://dom.spec.whatwg.org/#text) that is [decorative](#decorative), or
> - the test target is part of a [text node](https://dom.spec.whatwg.org/#text) that doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

> For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

Author: This would still express something in human language. I don't think this is a concern.

> For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

Reviewer 2: The text nodes would be "hel" and "lo". None of them express something in human language.

I see the second exception in the expectation as "except if the text node (in which the test target, i.e. the character is) doesn't express anything". 

Given that "hel" doesn't express anything in human language, each of the three characters would fall into that exception and all would pass.
That is,
```html
<span style="color: #FFE; background: #FFF;">hel</span>
```
is passing the rule (for each of the 3 test targets) because the text node does not express anything in human language. And similarly,
```html
<span style="color: #FFE; background: #FFF;">hel</span><span style="color: #FFD; background: #FFF;">lo</span>
```
passes the rule (for each of the 5 test targets) because none of the text nodes express anything in human language.

I do agree that it is not a big concern given that it is a edge case and will only result in passing a few extra things.

> For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).

Reviewer 3: I personally rather like the repetitiveness as it makes the conditions very explicit. I don't feel strongly about it though.

## Background

- [Understanding Success Criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html)
- [G18: Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G18)
- [G145: Ensuring that a contrast ratio of at least 3:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G145)
- [F83: Failure of Success Criterion 1.4.3 and 1.4.6 due to using background images that do not provide sufficient contrast with foreground text (or images of text)](https://www.w3.org/WAI/WCAG21/Techniques/failures/F83)
- [CSS Scoping Module Level 1 (Editor's Draft)](https://drafts.csswg.org/css-scoping/)

## Test Cases

### Passed

#### Passed Example 1

This dark grey text is on a white background.

```html
<p style="color: #333; background: #FFF;">

> 	Some text in a human language

Reviewer 2: Maybe using a text with more meaning than "this is a text" (because the font example later says that text used just for the sake of showing some text is not really text, so it could make sense that the other examples have text with a bit more "punch" in it that just being a text).

I suggest "I love ACT Rules!" (on all examples).

But do as you will, this is not a strong suggestion…

Author: It doesn't really matter.

</p>
```

#### Passed Example 2

This dark grey text is on a partially white gradient background.

```html
<p style="color: #333; background: linear-gradient(to right, #FFF, #00F); width: 500px;">
	Some text in a human language
</p>
```

#### Passed Example 3

This light grey text is on a partially black background image.

```html
<p
	style="color: #CCC; height:50px; padding-top:15px; background: #000 no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Passed Example 4

This text passes because the text shadow gives the text sufficient contrast.

```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>
```

#### Passed Example 5

This text only has a 3:1 contrast but it is a large (18 point) text so it still passes.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in a human language
</p>
```

#### Passed Example 6

> This text only has a 3:1 contrast but it is a large (14 points and bold) text so it still passes.

Reviewer 2: ```suggestion
This text only has a 3:1 contrast but it is a large (14 points and a font weight of 700 or more) text so it still passes.
```

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

> The purpose of the Helvetica letters is to show what the font looks like. The actual words are irrelevant and could be replaced with any other sentence that includes all letters of the alphabet. Because of that, the rule passes, despite the text not meeting the 4.5:1 color contrast requirement.

Reviewer 2: ```suggestion
The purpose of the Helvetica letters is to show what the font looks like. The actual words are irrelevant and could be replaced with any other sentence that includes all letters of the alphabet. Because of that, this is considered as non-text content (the meaning does not matter) and the rule passes, despite the text not meeting the 4.5:1 color contrast requirement.
```

**Note**: Because this is non-text content, [success criterion 1.4.11 Non-text Contrast](https://www.w3.org/TR/WCAG21/#non-text-contrast) requires font example to have a color contrast of 3:1.

```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```

#### Passed Example 8

This text has an insufficient contrast but it does not express anything in human language so it still passes.

```html
<p style="color: #000; background: #666;">
	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@
</p>
```

#### Passed Example 9

Even though the text is in a shadow DOM tree, the text is part of a child [text node](https://dom.spec.whatwg.org/#text) of an HTML element in the flat tree and has sufficient contrast

```html
<p style="color: #333; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

> 

Reviewer 2: Looking at the Expectation, there seems to be missing a Passed Example with decorative text.

Author: Passed example 7 has decorative text. We'll update the example description to make this clearer.

Reviewer 2: I disagree.

I think that the text in Passed Example 7 (the font example) does not meet the WCAG definition of "pure decoration" that is the exception to the SC:
> serving only an aesthetic purpose, providing no information, and having no functionality

The text in that example has the functionality of showing the font. This interpretation (it is not pure decoration) is seconded by the note on that example saying that it needs a 3:1 contrast to meet 1.4.11. I agree that it qualifies as a "Graphical Object" for 1.4.11:
> Graphical Objects: Parts of graphics required to understand the content, except when a particular presentation of graphics is essential to the information being conveyed.
(even if it's "text graphics").

In any case, if the text in that example is "pure decoration", the note and the reference to 1.4.11 need to go away. If it is not "pure decoration" (which I think), then a "pure decoration" example should be added.

Author: "functionality" as it is used in WCAG refers to things users can do on the page. Adding an item to a shopping cart is functionality.  See the definition of [functionality](https://www.w3.org/TR/WCAG21/#dfn-functionality). The font shape isn't "information" either, otherwise 1.3.1 would apply, information conveyed through presentation. 

Additionally, if this wasn't "purely decorative", it wouldn't be exempt under the 1.4.3 "incidental" bullet at all. Here's what 1.4.3's understanding doc says about it:

> For example, if random words are used to create a background and the words could be rearranged or substituted without changing meaning, then it would be decorative and would not need to meet this criterion.

This is precisely the same here. You can change the words without changing the meaning. That makes it decorative.

Reviewer 2: I agree that the text is decorative.
I do not understand why a decorative text is relevant for 1.4.11 (as pointed out by the note in Passed Example 7).
(it is not a UI component, and it is not a "Graphical Object" either; so 1.4.11 does not apply)

This is not important enough to block that PR…

Reviewer 4: Passed Example 7 is a little confusing. The first paragraph passes because it is using browser default styles. If the second paragraph is decorative, then it is an example of inapplicable.

Not sure how to split it as an example though. Perhaps a different example of each is needed.

Author: I'll put a foreground and background color on passed example 7 to clear this up, and add a separate example for browser defaults. That's a good point.

		color: rgba(90, 90, 90, 0.8);
		background-position: top 0 left 0;
		background-image: linear-gradient(90deg, transparent, transparent 3.3em, black 3.3em, black 6em);
		padding: 0 1em;
	}
</style>
<span id="backgroundSplit">
	Hello world
</span>
```

### Inapplicable

#### Inapplicable Example 1

This is invisible text.

```html
<p style="display: none">Some invisible text in English</p>
```

#### Inapplicable Example 2

The text is inapplicable because it is positioned off screen, hence not [visible](#visible).

```html
<p style="position:absolute; top: -999em">Some invisible text in English</p>
```

#### Inapplicable Example 3

This text has the same foreground and background colors, hence it is not [visible](#visible).

```html
<p style="color: white; background: white;">Some white on white text in English</p>
```

#### Inapplicable Example 4

This text is not the child of an HTML element.

```html
<svg>
	<text x="0" y="15">I love SVG!</text>
</svg>
```

#### Inapplicable Example 5

This is not a [text node](https://dom.spec.whatwg.org/#text) but an image.

```html
<p>
	<img scr="../test-assets/contrast/example.png" alt="example" />
</p>
```

#### Inapplicable Example 6

> This text is contained in a native button.

Reviewer 2: I think description needs to point out that `button` elements have an implicit role of `button` which inherits from `widget`.
+ link to def of the `button` mapping in the AAM (looks like this is wanted in #1063 and we should do the correct thing from the start if possible…)

```html
<p style="color: #333; background: linear-gradient(to right, #FFF, #00F); width: 500px;">
	Some text in a human language
</p>
```

#### Passed Example 3

This light grey text has a contrast ratio between 13:1 and 5:1 on the background image.

```html
<p
	style="color: #CCC; height:50px; padding-top:15px; background: #000 no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
>
	Black hole sun
</p>
```

#### Passed Example 4

This black text has a contrast ratio between 6.1:1 and 9:1 on grey background with white text shadow on it.

```html
<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
	Some text in a human language
</p>
```

#### Passed Example 5

This 18pt large black text has a contrast ratio of 3.6:1 on the grey background.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in a human language
</p>
```

#### Passed Example 6

This 14pt bold black text has a contrast ratio of 3.6:1 on the grey background.

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

> The first `p` element is has a contrast ratio of 21:1 (default black on white). The second `p` element contains Helvetica text which is decorative, because it does not convey information or provides functionality; it's purpose is to show the aesthetic of the Helvetica font.

Reviewer 2: ```suggestion
The first `p` element is has a contrast ratio of 21:1 (default black on white). The second `p` element contains Helvetica text which is decorative, because it does not convey information or provides functionality; its purpose is to show the aesthetic of the Helvetica font.
```
Damn, I missed that one on previous review 🤦‍♂ Sorry.

```html
<p style="color: #aaa; background: #fff;" id="p"></p>
<script>
	const shadowRoot = document.getElementById('p').attachShadow({ mode: 'open' })
	shadowRoot.textContent = 'Some text in English'
</script>
```

#### Failed Example 7

This semi-transparent grey text has a contrast ratio between 2.3:1 and 4.2:1 on the black and white background. The light grey text is compared to the white section of the background and the dark grey text is compared to the black section of the background.

```html
<style>
	#backgroundSplit {
		color: rgba(90, 90, 90, 0.8);
		background-position: top 0 left 0;
		background-image: linear-gradient(90deg, transparent, transparent 3.3em, black 3.3em, black 6em);
		padding: 0 1em;
	}
</style>
<span id="backgroundSplit">
	Hello world
</span>
```

### Inapplicable

#### Inapplicable Example 1

This text is not [visible][] because of `display: none`.

```html
<p style="display: none">Some invisible text in English</p>
```

#### Inapplicable Example 2

This text is not [visible][] because it is positioned off screen.

```html
<p style="position:absolute; top: -999em">Some invisible text in English</p>
```

#### Inapplicable Example 3

> This text is not [visible][] because the foreground color is the same as the background color.

Reviewer 4: Is this a WCAG allowable way to "hide" text? The moment a user applies their own styles, or selects content that includes it, the text would be visible.

Reviewer 3: If the assumption is "user styles can impact _any_ CSS" then we cannot ever say _anything_ about CSS. For example, hiding something by means of `display: none` would be just as "invalid" as hiding something by means of color; after all, it won't be hidden if overridden by user styles.

It's a slippery slope. If we need to account for the possibility of user styles, we need to be _very_ specific about which properties we expect users to override.

Reviewer 4: That's a fair comment. However, changing text and background colours is a very common accessibility change, eg. for high contrast or invert modes. Something that could unintentionally reveal such text. 

I was questioning the use of matching text and background colour as an acceptable way to "hide" text, because I don't think it is in keeping with the WCAG guidelines and was commonly used as a method to cheat SEO. There are more acceptable ways to visually hide text while keeping it part of the DOM, and this particular method breaks the minimal contrast guideline. I don't think it is inapplicable at all, and am wondering why it is included.

Reviewer 3: I definitely agree that using identical fore- and background colors as a means of hiding text is questionable. What I was trying to get at is that there's a subtle, but important, difference between this...

> This text should not be considered hidden because user styles can override text color.

...and this:

> This text should not be considered hidden because text color can be assumed to be overridden by user styles.

I'd be perfectly okay with not considering identical fore- and background colors as a means of hiding text for the rules where it makes sense, of course coupled with an assumption that explains _why_.

Author: That's a very good point Emma. Thanks for bringing this up. I see where you're coming from, custom styles / colors matter for 1.4.3, as the understanding doc explains:

> source: https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html
> It is a failure if no background color is specified when the text color is specified, because the user's default background color is unknown and cannot be evaluated for sufficient contrast.

I agree with you this is an accessibility problem. But I suspect that in most cases such texts should be treated as "purely decorative", which is an exception under 1.4.3. For the SEO example text for example, it isn't there for an aesthetic purpose, but it doesn't provide information or functionality. Not to the user anyway.

I'll add an assumption to the rule to cover this case. I think there are cases where a 1:1 contrast text fails 1.4.3, but my guess is that usually it doesn't.

Reviewer 4: I would not consider text the same colour as the background to be decorative, unless it was using a drop-shadow or similar to create a visible pattern of some kind. An Inapplicable Example that is clearly demonstrating decorative text would be a different matter.

https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html includes...
"This requirement applies to situations in which images of text were intended to be understood as text. Incidental text, such as in photographs that happen to include a street sign, are not included. Nor is text that for some reason is designed to be invisible to all viewers."

That is the only exception I have found any information about, and I'm not sure this is a good example of that.

Reviewer 4: Perhaps something like:

This text is not fully [visible][] because it is intended to be decorative rather than meaningful, so colour minimal contrast is not necessary.
```html
<div style="color: white; background: white; text-shadow: 2px 2px;" aria-hidden="true">Meaningless white on white text</div>
```

Reviewer 4: Or at least a note that this is not a recommended technique... and an additional example of decorative incidental meaningless text.

## Assumptions

- [Success criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) has exceptions for "incidental" text, which includes inactive user interface components and decorative texts. The rule assumes that [text nodes](https://dom.spec.whatwg.org/#text) that should be ignored are [disabled]() or hidden from assistive technologies. If this isn't the case, the rule may produce incorrect results.

- [Success criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) also has an exception for logos and brand names. Since logos and brand names are usually displayed through images to ensure correct rendering, this rule does not take logos or brand names into consideration. If a logo or brand name is included using [text nodes](https://dom.spec.whatwg.org/#text), this rule may produce incorrect results.

> - Text that has the same foreground and background color (a contrast ratio of 1:1) is assumed to be [purely decorative](). The text is assumed to not convey information that is meant for users. Hiding text in this way is sometimes used for search engine optimisation. Text with a contrast of 1:1 does not fail this rule, but may not satisfy [Success criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) if it is not decorative.

Reviewer 4: I'm not sure this is the correct wording to add. It isn't something to encourage.

  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag20:1.4.6: # Contrast (Enhanced)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

acknowledgements:
  authors:
    - Brian Bors
    - Kasper Isager
    - Wilco Fiers
---

## Applicability

Any [visible][] character in a [text node][] that is a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node][] is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:

- has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget); or
- is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled][]; or
- has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) and is [disabled][].

> **Note**: When the [foreground color](#foreground-colors-of-text) is the same as the [background color](#background-colors-of-text), the character is not [visible][], and so it does not need to be tested for contrast.

Reviewer 4: ```suggestion
**Note**: When the [foreground color](#foreground-colors-of-text) is the same as the [background color](#background-colors-of-text), this rule deems the character not [visible][], so it does not need to be tested for contrast.
```

This 18pt large black text has a contrast ratio of 3.6:1 on the gray background.

```html
<p style="color: #000; font-size:18pt; background: #666;">
	Some text in a human language
</p>
```

#### Passed Example 6

This 14pt bold black text has a contrast ratio of 3.6:1 on the gray background.

```html
<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
	Some text in English
</p>
```

#### Passed Example 7

The first `p` element is has a contrast ratio of 21:1 (default black on white). The second `p` element contains Helvetica text which is decorative, because it does not convey information or provides functionality; its purpose is to show the aesthetic of the Helvetica font.

**Note**: Because this is non-text content, [success criterion 1.4.11 Non-text Contrast](https://www.w3.org/TR/WCAG21/#non-text-contrast) requires font example to have a color contrast of 3:1.

```html
<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
	The quick brown fox jumps over the lazy dog.
</p>
```

#### Passed Example 8

> 

Reviewer 4: Is Passed Example 8 intended to be an example of something non-meaningful or decorative? If so, should it not be an Inapplicable Example?

Author: No, this is an exception mentioned in the expectation. This is because "decorative" is subjective, and the ACT rules format requires the applicability to be objective.

Reviewer 4: So passed example 7 is including text that can be considered decorative (visual only), and passed example 8 is including text that is not conveying meaning.

And in this instance the exceptions are passing examples rather than inapplicable examples because ... ?  The only reason I can determine is because the exception is in the expectation rather than the applicability. Is this consistent across rules? Should it be added to the rule guidance docs?

	shadowRoot.textContent = 'Some text in English'
</script>
```

#### Failed Example 7

This semi-transparent gray text has a contrast ratio between 2.3:1 and 4.2:1 on the black and white background. The light gray text is compared to the white section of the background and the dark gray text is compared to the black section of the background.

```html
<style>
	#backgroundSplit {
		color: rgba(90, 90, 90, 0.8);
		background-position: top 0 left 0;
		background-image: linear-gradient(90deg, transparent, transparent 3.3em, black 3.3em, black 6em);
		padding: 0 1em;
	}
</style>
<span id="backgroundSplit">
	Hello world
</span>
```

### Inapplicable

#### Inapplicable Example 1

This text is not [visible][] because of `display: none`.

```html
<p style="display: none">Some invisible text in English</p>
```

#### Inapplicable Example 2

This text is not [visible][] because it is positioned off screen.

```html
<p style="position:absolute; top: -999em">Some invisible text in English</p>
```

#### Inapplicable Example 3

This text is not [visible][] because the foreground color is the same as the background color.

```html

> <p style="color: white; background: white;" aria-hidden="true">Page last updated on 2020/01/10</p>

Reviewer 4: ```suggestion
<p style="color: white; background: white;" aria-hidden="true">Hidden text - U U D D L R L R B A S</p>
```

Reviewer 4: Page last updated is often useful information that should be visible. I've suggested a fun easter egg, that makes sense to be hidden in a findable way.

- [Success criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/TR/WCAG21/#contrast-minimum) also has an exception for logos and brand names. Since logos and brand names are usually displayed through images to ensure correct rendering, this rule does not take logos or brand names into consideration. If a logo or brand name is included using [text nodes](https://dom.spec.whatwg.org/#text), this rule may produce incorrect results.

> - Text that has the same foreground and background color (a contrast ratio of 1:1) is not considered to be "visual presentation of text", making it inapplicable to the success criterion. Text hidden in this way can still cause accessibility issues under other success criteria, depending on the content.

Reviewer 2: Should this be in "Accessibility Support" rather than in "Assumption"?

Author: Nope. Acc support is about techniques not working consistently between assistive tech. This is an interpretation of the success criterion. That belongs in assumptions.
